torch/distributed/_composable/fsdp/_fsdp_api.py

```diff
@@ -59,6 +59,31 @@ def __post_init__(self):
             object.__setattr__(self, "reduce_dtype", None)
 
 
+@dataclass
+class OffloadPolicy:
+    """This base class represents the policy of no offloading."""
+
+
+@dataclass
+class CPUOffloadPolicy(OffloadPolicy):
```

Comment:
Naive question: why is this all or nothing, i.e. why doesn't the user have control over parameters, gradients, optimizer states separately

Reply:
In my view, the main answer is that it is the simplest schedule that also happens to minimize the H2D/D2H copy bandwidth usage.

However, given that currently the optimizer step computation itself is the bottleneck and not the copying, I think it is fair to consider other schedules.

In this PR, FSDP is only responsible for (1) keeping sharded parameters on CPU, (2) copying them H2D as needed for forward/backward computation, and (3) copying sharded gradients D2H after reduce-scatter in backward. FSDP does not manipulate the optimizer explicitly.

Therefore, from FSDP's perspective, there are only two cases to discuss:
1. Keep sharded parameters on GPU but offload sharded gradients D2H to CPU
2. Keep sharded parameters on CPU but keep sharded gradients on GPU

Anything related to the optimizer state can be handled in the optimizer outside of FSDP. Note also that these two cases are only relevant to trying to run `optim.step()` on GPU, not CPU. If `optim.step()` is running on CPU, then the existing schedule is already optimal.

Some of my concerns with these two cases:
- Both of these cases would require the user to externally add logic such that parameters, gradients, and optimizer states are all on the same device at the time of `optim.step()` (i.e. this would not work out of the box with core optimizers). For example for case 1, the optimizer would need to copy a sharded gradient back to GPU, run the optimizer step for the parameter, and offload that gradient back to CPU. This could possibly be handled with optimizer hooks, though I wonder who is responsible for defining and registering these hooks. The efficiency for this kind of pattern should be evaluated too (e.g. for all parameters: copy H2D -> small computation -> copy D2H).
- The optimizer states contribute 2x memory compared to parameters or gradients each. For memory savings, we prefer for the optimizer states to be on CPU. However, we said that these two cases are only relevant when we run optimizer on _GPU_. This means either less memory savings or extra logic to also copy H2D/D2H optimizer states around `optim.step()`.
- PyTorch today does not let you have a tensor that is on a different device than its `.grad`. This means to support either of the two cases, we need to do something "extra" to save the parameters and gradients, while exposing them still to the user.
- Since FSDP is module-level, e.g. applied to each transformer block, this CPU offloading can be applied to one transformer block but not another (for example). This still offers some level of tuning in terms of how much they want to offload (albeit, not in the same way as the two aforementioned cases).